pkg/cachemanager/cachemanager.go

```diff
@@ -140,47 +157,77 @@ func (c *CacheManager) UpsertSource(ctx context.Context, sourceKey aggregator.Ke
-
-	var innerError error
+	var err error
 	c.watchedSet.Replace(newWatchSet, func() {
 		// *Note the following steps are not transactional with respect to admission control
 
 		// Important: dynamic watches update must happen *after* updating our watchSet.
 		// Otherwise, the sync controller will drop events for the newly watched kinds.
-		innerError = c.registrar.ReplaceWatch(ctx, newWatchSet.Items())
+		err = c.registrar.ReplaceWatch(ctx, newWatchSet.Items())
 	})
 
-	return innerError
+	return err
 }
 
-// removeStaleExpectations stops tracking data for any resources that are no longer watched.
-func (c *CacheManager) removeStaleExpectations(stale *watch.Set) {
-	for _, gvk := range stale.Items() {
-		c.tracker.CancelData(gvk)
+// interpret if the err received is general or whether it is specific to the provided GVKs.
+// also returns whether there were any errors removing watches.
+func interpretErr(e error, gvks []schema.GroupVersionKind) (bool, []schema.GroupVersionKind, bool) {
+	if e == nil {
+		return false, nil, false
 	}
+
+	f := watch.NewErrorList()
+	if !errors.As(e, &f) || f.HasGeneralErr() {
+		return true, nil, false
+	}
+
+	failedGvksToAdd := watch.NewSet()
+	failedGvksToAdd.Add(f.FailingGVKsToAdd()...)
+	sourceGVKSet := watch.NewSet()
+	sourceGVKSet.Add(gvks...)
+
+	common := failedGvksToAdd.Intersection(sourceGVKSet)
+	if common.Size() > 0 {
+		return false, common.Items(), f.HasFailingGVKsToRemove()
+	}
+
+	// this error is not about the gvks in this request
+	// but we still log it for visibility
+	log.V(logging.DebugLevel).Info("encountered unrelated error when replacing watch set", "error", e)
+	return false, nil, f.HasFailingGVKsToRemove()
 }
 
 // RemoveSource removes the watches of the GVKs for a given aggregator.Key. Callers are responsible for retrying on error.
 func (c *CacheManager) RemoveSource(ctx context.Context, sourceKey aggregator.Key) error {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	if err := c.gvksToSync.Remove(sourceKey); err != nil {
-		return fmt.Errorf("internal error removing source: %w", err)
+	c.gvksToSync.Remove(sourceKey)
+	err := c.replaceWatchSet(ctx)
+	general, _, danglingWatches := interpretErr(err, []schema.GroupVersionKind{})
+	if general {
+		return fmt.Errorf("error establishing watches: %w", err)
 	}
 
-	if err := c.replaceWatchSet(ctx); err != nil {
-		return fmt.Errorf("error removing watches for source %v: %w", sourceKey, err)
-	}
+	// watch removal retries for any dangling watches will happen async in manageCache()
+	c.handleDanglingWatches(danglingWatches)
 
 	return nil
 }
 
+// Mark if there are any dangling watches that need to be retried to be removed.
+// assumes caller has lock.
+func (c *CacheManager) handleDanglingWatches(danglingWatches bool) {
+	// if there were dangling watches previously but not anymore,
+	// mark the data cache as needing a wipe so we don't track unwanted resources.
+	if c.danglingWatches && !danglingWatches {
+		c.forceWipe = true
```

Comment:
We also need to set forceWipe to true if there has been any removal of GVKs, I think that logic got lost somewhere.

Reply:
I'm not sure I follow why we should set `forceWipe` if we remove GVKs. Please let me know if I am missing something obvious.

As written, this chunk of code accounts for the scenario where we had dangling watches to be removed in the past and we no longer have that. (One  possible presumption for such a case is that removal of said dangling was successful)

In the happy path, when we do remove the watches for GVKs without an error, we already have that accounting in place with `c.gvksToDeleteFromCache`:

https://github.com/open-policy-agent/gatekeeper/blob/2055dccad326625e292acf65717e6ea22bee00d6/pkg/cachemanager/cachemanager.go#L147

which gets taken into account in `wipeCacheIfNeeded` every time we run `manageCache`.

Reply:
Ah, we probably shouldn't have two mechanisms to do the same thing. I suggest altering  `HasRemoveGVKErrors()` to return the GVKs with removal errors (updating function name appropriately), and adding those to `gvksToDeleteFromCache`.

As a bonus, that narrows down which GVKs need a wipe once we have the iterable cache.

An alternative would be to move everything to `forceWipe`